### PR TITLE
Merge correlation IDs from event context and event data

### LINKED ISSUE #668: Duplicate correlation IDs in `gs.up.receive`
#### Summary

The correlation IDs in the `gs.up.receive` event are duplicate.

#### Steps to Reproduce

1. Subscribe to events, i.e. `ttn-lw-cli events subscribe --gateway-id ...`

#### What do you see now?

For example:

```json
{
  "name": "gs.up.receive",
  "time": "2019-05-10T08:48:36.397803Z",
  "identifiers": [
    {
      "gateway_ids": {
        "gateway_id": "tektelic-micro",
        "eui": "647FDAFFFE0059AB"
      }
    }
  ],
  "data": {
    "@type": "type.googleapis.com/ttn.lorawan.v3.UplinkMessage",
    "raw_payload": "QKogASYAkdgBU2NToyoatD/C",
    "settings": {
      "data_rate": {
        "lora": {
          "bandwidth": 125000,
          "spreading_factor": 7
        }
      },
      "coding_rate": "4/5",
      "frequency": "867900000",
      "timestamp": 3375161699
    },
    "rx_metadata": [
      {
        "gateway_ids": {
          "gateway_id": "tektelic-micro",
          "eui": "647FDAFFFE0059AB"
        },
        "timestamp": 3375161699,
        "rssi": 2,
        "snr": 10,
        "uplink_token": "ChwKGgoOdGVrdGVsaWMtbWljcm8SCGR/2v/+AFmrEOPCs8kM"
      }
    ],
    "received_at": "2019-05-10T08:48:36.397329Z",
    "correlation_ids": [
      "gs:conn:01DAGEFV4TBH5K9AVAX86NPDXM",
      "gs:uplink:01DAGEW31DQZQNSCQPJ28KGFMF"
    ],
    "gateway_channel_index": 4
  },
  "correlation_ids": [
    "gs:conn:01DAGEFV4TBH5K9AVAX86NPDXM",
    "gs:uplink:01DAGEW31DQZQNSCQPJ28KGFMF",
    "gs:conn:01DAGEFV4TBH5K9AVAX86NPDXM",
    "gs:uplink:01DAGEW31DQZQNSCQPJ28KGFMF"
  ],
  "origin": "Johans-MacBook-Pro.local"
}
```

#### What do you want to see instead?

No duplicates in `correlation_ids`

#### How do you propose to implement this?

The correlation IDs probably get added twice.

#### Can you do this yourself and submit a Pull Request?

Yes

### PR DESCRIPTION
**Pull request by pgalic96**
#### Summary

Closes #668 

#### Changes

- Use MergeSort to merge correlation ids from context and from the payload
- Add test for duplicate correlation ids

#### Notes for Reviewers

Implemented suggestion by @htdvisser to use the MergeSort algorithm already present in `correlation_context.go`.

### CODE REVIEW

File: pkg/events/events_test.go [hunk 1]
```diff
@@ -65,7 +82,7 @@ func TestEvents(t *testing.T) {
 	pubsub.Subscribe("test.*", handler)
 	pubsub.Subscribe("test.*", handler) // second time should not matter
 
-	evt := events.New(ctx, "test.evt1", nil, "hello")
+	evt := events.New(ctx, "test.evt1", nil, "")
```
Comment (rvolosatovs) [inline pkg/events/events_test.go]: Unrelated change

**Comment by rvolosatovs** [comment]
@pgalic96 please fix the PR title and remove `Release Notes` section, if not applicable

**Comment by rvolosatovs** [comment]
Also, set milestone on PRs in the future

File: pkg/events/events.go [hunk 2]
```diff
@@ -145,7 +145,10 @@ func New(ctx context.Context, name string, identifiers CombinedIdentifiers, data
 		data: data,
 	}
 	if data, ok := data.(interface{ GetCorrelationIDs() []string }); ok {
-		evt.innerEvent.CorrelationIDs = append(evt.innerEvent.CorrelationIDs, data.GetCorrelationIDs()...)
+		if cids := data.GetCorrelationIDs(); len(cids) > 0 {
+			ctx := ContextWithCorrelationID(ctx, cids...)
```
Comment (rvolosatovs) [inline pkg/events/events.go]: Please revert this change, having a dedicated function was good, but it should not be exported.

File: pkg/events/events_test.go [hunk 3]
```diff
@@ -32,6 +32,19 @@ type wrappedEvent struct {
...
+func (testData) GetCorrelationIDs() []string {
+	return []string{"TestNew"}
+}
+
+func TestNew(t *testing.T) {
+	a := assertions.New(t)
+	events.IncludeCaller = true
+	ctx := events.ContextWithCorrelationID(test.Context(), t.Name())
+	evt := events.New(ctx, "as.up.receive", nil, testData{})
+	a.So(evt.CorrelationIDs(), should.HaveLength, 1)
```
Comment (rvolosatovs) [inline pkg/events/events_test.go]:
```suggestion
	a.So(evt.CorrelationIDs(), should.Resemble, []string{"TestNew"})
```

File: pkg/events/correlation_context.go [hunk 4]
```diff
@@ -33,9 +33,28 @@ func ContextWithCorrelationID(ctx context.Context, cids ...string) context.Conte
 	if !ok {
 		return context.WithValue(ctx, correlationKey{}, cids)
 	}
+	merged := mergeCorrelationIDs(existing, cids)
+	return context.WithValue(ctx, correlationKey{}, merged)
```
Comment (rvolosatovs) [inline pkg/events/correlation_context.go]: Nitpick; you can just inline the `mergeCorrelationIDs` call here

File: pkg/events/correlation_context.go [hunk 5]
```diff
@@ -33,9 +33,28 @@ func ContextWithCorrelationID(ctx context.Context, cids ...string) context.Conte
...
+}
+
+// NewCorrelationID returns a new random correlation ID.
+func NewCorrelationID() string {
+	return ulid.MustNew(ulid.Now(), rand.Reader).String()
+}
 
-	// Since existing and cids are sorted, merging them will result in a sorted slice.
-	// See https://en.wikipedia.org/wiki/Merge_sort
+// mergeCorrelationIDs returns sorted array of correlation ids from context and payload
```
Comment (rvolosatovs) [inline pkg/events/correlation_context.go]: 1. This function returns a slice, not an array
2. This function merges 2 sorted string slices, it has nothing to do with correlation IDs, context or payload

File: pkg/events/correlation_context.go [hunk 6]
```diff
@@ -33,9 +33,28 @@ func ContextWithCorrelationID(ctx context.Context, cids ...string) context.Conte
...
+// Since existing and cids are sorted, merging them will result in a sorted slice.
+// See https://en.wikipedia.org/wiki/Merge_sort
+func mergeCorrelationIDs(existing, cids []string) []string {
```
Comment (rvolosatovs) [inline pkg/events/correlation_context.go]:
```suggestion
func mergeStrings(a, b []string) []string {
```

File: pkg/events/correlation_context.go [hunk 7]
```diff
@@ -33,9 +33,28 @@ func ContextWithCorrelationID(ctx context.Context, cids ...string) context.Conte
...
 	merged := make([]string, 0, len(existing)+len(cids))
```
Comment (rvolosatovs) [inline pkg/events/correlation_context.go]:
```suggestion
	merged := make([]string, 0, len(a)+len(b))
```

File: pkg/events/correlation_context.go [hunk 8]
```diff
@@ -33,9 +33,28 @@ func ContextWithCorrelationID(ctx context.Context, cids ...string) context.Conte
...
 	a, b := existing, cids
```
Comment (rvolosatovs) [inline pkg/events/correlation_context.go]:
```suggestion
```

File: pkg/events/events_test.go [hunk 9]
```diff
@@ -32,6 +32,20 @@ type wrappedEvent struct {
...
 
+type testData struct{}
+
+func (testData) GetCorrelationIDs() []string {
+	return []string{"TestNew"}
+}
+
+func TestNew(t *testing.T) {
+	a := assertions.New(t)
+	events.IncludeCaller = true
```
Comment (rvolosatovs) [inline pkg/events/events_test.go]: Why?
